### ISSUE TITLE
Add methods to convert between hexadecimal strings and bytes

### ISSUE THREAD

**Issue by GSPP** [dotnet/corefx#10013]
It is quite common to need to convert bytes to hex strings and back. The .NET Framework does not have an API to do that. Look at what lengths people have gone to to solve this problem for them: [How do you convert Byte Array to Hexadecimal String, and vice versa?](http://stackoverflow.com/questions/311165/how-do-you-convert-byte-array-to-hexadecimal-string-and-vice-versa) Many of those solutions are somehow bad (e.g. slow or broken or not validating).

`BitConverter.ToString(byte[])` generates strings of a practically useless format: `BitConverter.ToString(new byte[] { 0x00, 0x01, 0xAA, 0xFF })` => `00-01-AA-FF`. I don't understand how this decision was made. This seems to be a special solution for some specific purpose (debug output?). Using this API requires `Replace(str, "-", "")` which is ugly and slow. Also, this API is not discoverable because it does not have "hex" in it's name.

I propose adding an API that does the following:
* Supports char[] and string
* Supports specifying the output buffer to support zero-allocation scenarios
* Supports upper- and lowercase
* Fast
* Validates input
* *No* support for any kind of hex prefix (such as "0x"), whitespace or digit separator (such as "-" or "\n"). This is out of scope.

The use case for char[] in- and outputs is low-allocation scenarios.

The API could look like this:

```
static class HexConvert {
 byte[] ToBytes(string input, int start, int count, HexFormat hexFormat);
 void ToBytes(string input, int start, int count, ArraySegment<byte> outputBuffer, HexFormat hexFormat);
 //Repeat the same for ArraySegment<char> as input

 string ToString(ArraySegment<byte> bytes);
 char[] ToChars(ArraySegment<byte> bytes);
 void ToChars(ArraySegment<byte> bytes, ArraySegment<char> outputBuffer);

 //Plus obvious convenience overloads

 int HexDigitToInteger(char val); //0-9A-F => 0-15.
 char IntegerToHexDigit(int value, HexFormat hexFormat); //0-15 => 0-9A-F
}

[Flags]
enum HexFormat { UseDefaults = 0, Lowercase = 1 } //extensible
```

A possible extension would be to support `Stream` and `IEnumerable<byte>` as sources and destinations. I would consider this out of scope.

I feel that converting hexadecimal strings is an important and frequent enough thing that this should be supported well. Of course, a "light" version of this proposal is also feasible (just support for string and byte[] with no user-provided output buffer). Not sure if anyone wants/needs lowercase hex, I don't but others might.

This seems cheap enough to implement. Testing should be easily automatable. I don't see any particular usability problems.

**Comment by AlexGhiondea** [comment]
@GSPP how about augmenting the BitConverter type with these new methods?

I feel like that would be a better place for these since that is already a type people use for converting between various formats.

**Comment by AlexGhiondea** [comment]
There are at least a couple of other issues tracking something like this:
https://github.com/dotnet/corefx/issues/1647
https://github.com/dotnet/corefx/issues/4441

Let's track it with a single issue (this one).

It does look like this is something for which there is interest in implementing.

@GSPP, @GrabYourPitchforks @KrzysztofCwalina is anyone of you interested in writing up a formal API proposal for this?

**Comment by CodesInChaos** [comment]
If the slices API lands before this, I'd go with something like:

```
// all instance methods are thread-safe
public abstract class BinaryEncoding
{
	protected abstract void ToBytesOverride(ReadOnlySpan<char> sourceChars, Span<byte> destinationBytes);
	protected abstract int GetByteLengthOverride(int stringLength);
	protected abstract void ToCharsOverride(ReadOnlySpan<byte> sourceBytes, Span<char> destinationChars);
	protected abstract int GetStringLengthOverride(int byteLength);

	public byte[] ToBytes(ReadOnlySpan<char> str);
	public void ToBytes(ReadOnlySpan<char> sourceChars, Span<byte> destinationBytes);
	public int GetByteLength(int stringLength);
	
	public string ToString(ReadOnlySpan<byte> bytes);
	public char[] ToChars(ReadOnlySpan<byte> bytes);
	public void ToChars(ReadOnlySpan<byte> sourceBytes, Span<char> destinationChars);
	public int GetStringLength(int byteLength);
}

public class HexEncoding
{
	public static HexEncoding UpperCase { get; }
	public static HexEncoding LowerCase { get; }
}
```

perhaps add some additional overload with `byte[]`, `char[]`, `string` inputs.

**Comment by karelz** [comment]
We don't take any new API additions if they would be obsolete when `Span` lands.
Marking as blocked on `Span` - @AlexGhiondea feel free to change it if you think I am wrong.

**Comment by GSPP** [comment]
Span has landed so this can be unblocked.

**Comment by GrabYourPitchforks** [comment]
I propose a slightly different API, instead putting this on the `Convert` class near the existing base64 APIs. That would help discoverability, and it gives us an existing template for the recommended API shape.

```cs
namespace System
{
    // NEW methods on EXISTING type
    public static class Convert
    {
        public static byte[] FromHexString(ReadOnlySpan<char> chars); // doesn't match an existing overload of FromBase64String, but could be useful
        public static byte[] FromHexString(string s);

        public static string ToHexString(byte[] inArray, HexFormattingOptions options = default);
        public static string ToHexString(byte[] inArray, int offset, int length, HexFormattingOptions options = default);
        public static string ToHexString(ReadOnlySpan<byte> bytes, HexFormattingOptions options = default);
    }

    [Flags]
    public enum HexFormattingOptions
    {
        None = 0,
        InsertLineBreaks = 1,
        Lowercase = 2,
    }
}
```

It also makes sense IMO to have `OperationStatus`-based APIs. Those APIs could be done in the corefxlab repo to live alongside the existing `Base64Encoder` and `Base64Decoder` types in that repo. Then all of the `OperationStatus`-based base64 and hex APIs can be reviewed at once if we decide to move them into corefx.

**Comment by vcsjones** [comment]
To give some parity with the `*CharArray` from base64, I would consider a few other methods:

```csharp
public static class Convert
{
    public bool TryFromHexString(ReadOnlySpan<char> chars, Span<byte> buffer, out int bytesWritten);
    public bool TryFromHexString(string s, Span<byte> buffer, out int bytesWritten);
    public bool TryToHexString(ReadOnlySpan<byte> bytes, Span<char> buffer, out int charsWritten, HexFormattingOptions options = default);
}
```

The rationale for this is in my typical usage I expect a hex string to be short (think a SHA digest) which tops out at 64 bytes right now.

I'm not strongly tied to these, but worth considering during review.

**Comment by GrabYourPitchforks** [comment]
@vcsjones I envision `HexFormattingOptions.InsertLineBreaks` as inserting every 72 chars, just like base64 does today. And parsing would allow all of uppercase / lowercase / whitespace, hence no options parameter.

**Comment by GSPP** [comment]
I'd *not* put it on `Convert`. My argument for that is this: Imagine there were 10 different such formats (Hex, base64, base85, ...). Then, we clearly would like each of them to be their own static class. Anything else would make for a very crowded API. Now we don't have 10 formats but from this thought experiment we can see that creating a separate class is the better option. Especially, since more formats might be added in the future. For example, Bitcoin uses a base85 format which is more efficient. Who knows if additional formats might be desirable in the future.

Should this API accept any other input besides spans? It could accept arrays, strings, `ArraySegment` and even pointers. In my opinion it should *only* take spans. That would make for the cleanest and least crowded API. Anything else is easily converted to span. In particular I don't want to see overloads with index and count parameters. Thank god we no longer need those!

We certainly need methods that do not throw in the error case. My original proposal missed those.

> `InsertLineBreaks` as inserting every 72 chars

Is this a widely known standard format? Is it only for base64 or also for hex?

<hr>

New API proposal:

```
namespace System
{
    public static class HexConvert
    {
        //convert to bytes, throwing
        public static byte[] GetBytes(ReadOnlySpan<char> chars, HexFormattingOptions formattingOptions = default);
        public static void GetBytes(ReadOnlySpan<char> chars, Span<byte> output, out int bytesWritten, HexFormattingOptions formattingOptions = default);

        //convert to bytes, try
        public static byte[] TryGetBytes(ReadOnlySpan<char> chars, HexFormattingOptions formattingOptions = default); //can return null
        public static bool TryGetBytes(ReadOnlySpan<char> chars, Span<byte> output, out int bytesWritten, HexFormattingOptions formattingOptions = default);

         //convert to chars
         string GetString(ReadOnlySpan<byte> bytes);
         char[] GetChars(ReadOnlySpan<byte> bytes);

         //plus non-throwing overloads
         //plus overloads with output buffer
         //plus overloads without HexFormattingOptions (I was lazy so I made them default arguments)
    }

    [Flags]
    public enum HexFormattingOptions
    {
        None = 0,
        Lowercase = 1,
    }
}
```

I chose the name prefix to be "Get" so that "TryGet" is fitting. This succinct naming would not be possible when placing these methods in `Convert`.

**Comment by AlexeiScherbakov** [comment]
There are many variants of API. It can be used in web,cryptography and many other places. Some hex conversion is used in json for writing non-ASCII chars - \u0000 - \uFFFF

I think that .NET Core must provide optimized variants for each purpose

**Writing API**
1. It can be uppercase/lowercase
2. It can use separators or not use
3. Separators can be single char or multi (char[] or string)
4. It can create new instance of string and char[] or use Span<char>, StringBuilder, TextWriter or Stream  as output 
5. Source can be Span<byte> or types like,(s)byte,(u)short,char,(u)int,(u)long
6. If source is Span<byte> function can take entire span or it part (for direct byte[] calls)

All above give us more than 100 function declarations. If it will have byte[] inputs - it take much more

For example just writing API for ReadOnlySpan<byte> as source, with all kinds of separators, and only lowercase give us 6 optimized variants
```
public static string ToHexStringLowerCase(this ReadOnlySpan<byte> array);
public static string ToHexStringLowerCase(this ReadOnlySpan<byte> array, char separator);
public static string ToHexStringLowerCase(this ReadOnlySpan<byte> array, ReadOnlySpan<char> separator);
public static string ToHexStringLowerCase(this ReadOnlySpan<byte> array, int offset, int count);
public static string ToHexStringLowerCase(this ReadOnlySpan<byte> array, int offset, int count, char separator);
public static string ToHexStringLowerCase(this ReadOnlySpan<byte> array, int offset, int count, ReadOnlySpan<char> separator);
```

This is bad idea to place it in existing places, I think, that better to place this API in external netcoreapp2.1+netstandard2.1 package. Something like System.Text.Hex

**Comment by AlexeiScherbakov** [comment]
Also, as @vcsjones mentioned for short strings, we need some different allocation mechanism, to avoid dirty hacks for reducing allocations count: (https://docs.microsoft.com/en-us/dotnet/csharp/how-to/modify-string-contents#unsafe-modifications-to-string)

```
public static unsafe string ToHexStringLowerCase(this byte[] array)
{
	string ret = new string('\0', array.Length * 2);
	
	fixed(char* strPtr=ret)
	{
		char* ptr = strPtr;
		for (int i = 0; i < array.Length; i++)
		{
			byte bytic = array[i];
			*ptr = HexChars.HexDigitsLowerCase[(bytic >> 4) & 0xF];
			ptr++;
			*ptr = HexChars.HexDigitsLowerCase[bytic & 0xF];
			ptr++;
		}
	}
	return ret;
}
```
For example in this case we can allocate string memory (x2 of byte array length) and write to it, without copy contructor.

**Comment by tkp1n** [comment]
I consolidated the different proposals and comments made in this discussion and came up with the following API. I also provide a sample implementation (internally utilizing AVX2, SSE4.1, SSSE3, SSE2 or Scalar/Lookup with the following [performance characteristics](https://github.com/tkp1n/HexMate#performance-characteristics)) of the API to experiment with via a seperate GitHub repository and nuget package. 

#### Sample implementation to play arround with
https://github.com/tkp1n/HexMate
https://www.nuget.org/packages/HexMate

#### Adressed proposals and comments:
* Parity with base64 APIs on `System.Convert` (@GrabYourPitchforks, https://github.com/dotnet/corefx/issues/10013#issuecomment-463419321 and @vcsjones, https://github.com/dotnet/corefx/issues/10013#issuecomment-463452088)
* `OperationStatus`-based APIs as currently in `System.Buffers.Text.Base64` (@GrabYourPitchforks, https://github.com/dotnet/corefx/issues/10013#issuecomment-463419321)
* Allocation-free or -friendly implementation is possible (see repo above, @AlexeiScherbakov, https://github.com/dotnet/corefx/issues/10013#issuecomment-464481397)

#### Open questions:
* Would it make sense to add APIs where one could choose a separator at the cost of additional overloads and deviating from base64 API parity? (@AlexeiScherbakov, https://github.com/dotnet/corefx/issues/10013#issuecomment-464480043)
* Why should `InsertLineBreaks` add a new-line every 72 characters for hex but [every 76 for base64](https://github.com/dotnet/corefx/blob/4ddf96fe04a6befe8814651a9933258379e3f4f4/src/Common/src/CoreLib/System/Convert.cs#L136)? (@GrabYourPitchforks, https://github.com/dotnet/corefx/issues/10013#issuecomment-463492630)
* Should the parsing routines accept `-`-separated input to allow processing strings returned by `BitConverter.ToString(byte[])`?
* Would `GetMaxDecodedFromUtf8Length` and `GetMaxEncodedToUtf8Length` make sense for `Hex` as the calculation is much simpler as with base64 where those methods are provided?

### API proposal

```csharp
namespace System
{
    // NEW type
    [Flags]
    public enum HexFormattingOptions
    {
        None = 0,
        InsertLineBreaks = 1,
        Lowercase = 2
    }

    // NEW methods on EXISTING type
    public static class Convert
    {
        // Decode from chars
        public static byte[] FromHexCharArray(char[] inArray, int offset, int length) => throw null;
        public static bool TryFromHexChars(ReadOnlySpan<char> chars, Span<byte> bytes, out int bytesWritten) => throw null;

        // Decode from strings
        public static byte[] FromHexString(string s) => throw null;
        public static byte[] FromHexString(ReadOnlySpan<char> chars) => throw null;
        public static bool TryFromHexString(string s, Span<byte> bytes, out int bytesWritten) => throw null;

        // Encode to chars
        public static int ToHexCharArray(byte[] inArray, int offsetIn, int length, char[] outArray, int offsetOut, HexFormattingOptions options = default) => throw null;
        public static bool TryToHexChars(ReadOnlySpan<byte> bytes, Span<char> chars, out int charsWritten, HexFormattingOptions options = default) => throw null;

        // Encode to strings
        public static string ToHexString(byte[] inArray, HexFormattingOptions options = default) => throw null;
        public static string ToHexString(byte[] inArray, int offset, int length, HexFormattingOptions options = default) => throw null;
        public static string ToHexString(ReadOnlySpan<byte> bytes, HexFormattingOptions options = default) => throw null;
    }
}

namespace System.Buffers.Text
{
    // NEW type
    public static class Hex
    {
        // Decode
        public static OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> utf8, Span<byte> bytes, out int bytesConsumed, out int bytesWritten, bool isFinalBlock = true) => throw null;
        public static OperationStatus DecodeFromUtf8InPlace(Span<byte> buffer, out int bytesWritten) => throw null;
        
        // Encode
        public static OperationStatus EncodeToUtf8(ReadOnlySpan<byte> bytes, Span<byte> utf8, out int bytesConsumed, out int bytesWritten, bool isFinalBlock = true) => throw null;
        public static OperationStatus EncodeToUtf8InPlace(Span<byte> buffer, int dataLength, out int bytesWritten) => throw null;
    }
}
```

**Comment by GSPP** [comment]
I think we should demonstrate a need for inserting line breaks first. I am not aware of any common scenario where this would be used. If this feature is added then the line length must be variable. I cannot imagine any fixed line length to be appropriate.

> Should the parsing routines accept --separated input

By default, the parsing routines should be maximally strict. This means *no* support for whitespace or any separator. When we parse data we generally want to be as strict as possible in order to find bugs in the originating systems and in order to not accidentally accept invalid data.

I see no need for supporting a separator. This is a rare need. An easy workaround exists at a performance cost: `str.Replace("-", "")`.

**Comment by terrajobst** [comment]
@GrabYourPitchforks, it seems we should survey what internal implementations we have and what all the places are where we expose something similar. It seems new proposals were added after you marked it as read for review.

**Comment by bugproof** [comment]
Really looking forward to seeing this. I benchmarked @tkp1n implementation and it's faster than everything else in .NET Core.

**Comment by tkstanczak** [comment]
[DebuggerStepThrough]
     public static string ByteArrayToHexViaLookup32Safe(byte[] bytes, bool withZeroX)
     {
         if (bytes.Length == 0)
         {
             return withZeroX ? "0x" : "";
         }
         
         int length = bytes.Length * 2 + (withZeroX ? 2 : 0);
         StateSmall stateToPass = new StateSmall(bytes, withZeroX);
         return string.Create(length, stateToPass, (chars, state) =>
         {
             int offset0x = 0;
             if (state.WithZeroX)
             {
                 chars[0] = '0';
                 chars[1] = 'x';
                 offset0x += 2;
             }

             Span<uint> charsAsInts = MemoryMarshal.Cast<char, uint>(chars.Slice(offset0x));
             int targetLength = state.Bytes.Length;
             for (int i = 0; i < targetLength; i += 1)
             {
                 uint val = Lookup32[state.Bytes[i]];
                 charsAsInts[i] = val;
             }
         });
     }

**Comment by tkstanczak** [comment]
```
[DebuggerStepThrough]
     public static string ByteArrayToHexViaLookup32Safe(byte[] bytes, bool withZeroX)
     {
         if (bytes.Length == 0)
         {
             return withZeroX ? "0x" : "";
         }
         
         int length = bytes.Length * 2 + (withZeroX ? 2 : 0);
         StateSmall stateToPass = new StateSmall(bytes, withZeroX);
         return string.Create(length, stateToPass, (chars, state) =>
         {
             int offset0x = 0;
             if (state.WithZeroX)
             {
                 chars[0] = '0';
                 chars[1] = 'x';
                 offset0x += 2;
             }

             Span<uint> charsAsInts = MemoryMarshal.Cast<char, uint>(chars.Slice(offset0x));
             int targetLength = state.Bytes.Length;
             for (int i = 0; i < targetLength; i += 1)
             {
                 uint val = Lookup32[state.Bytes[i]];
                 charsAsInts[i] = val;
             }
         });
     }
```

**Comment by tkstanczak** [comment]
It is possible that the additional checks added by API conformance make it slower but have a look at the results here:
|     Method |     Mean |     Error |    StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------- |---------:|----------:|----------:|------:|--------:|-------:|------:|------:|----------:|
| SafeLookup | 17.77 ns | 0.4267 ns | 0.6517 ns |  0.51 |    0.02 | 0.0119 |     - |     - |      56 B |
|   HexMateA | 34.28 ns | 0.9541 ns | 1.0209 ns |  1.00 |    0.00 | 0.0119 |     - |     - |      56 B |
